src/testing/observable_independence.py

```diff
@@ -0,0 +1,323 @@
+        attained using a permuted version of `x2_array` to predict `x1_array`,
+        given `z_array` if it was not None.
+    """
+    # Validate argument type and lengths
+    _ensure_is_array(x1_array, 'x1_array')
+    _ensure_is_array(x2_array, 'x2_array')
+    if z_array is not None:
+        _ensure_is_array(z_array, 'z_array')
+    _check_array_lengths(x1_array, x2_array, array_3=z_array)
+
+    # Set a random seed for reproducibility
+    if seed is not None:
+        np.random.seed(seed)
+
+    # Determine how to create the predictors for the permutation test, based
+    # on whether we want a marginal independence test (i.e. z_array = None)
+    # or a conditional independence test (isinstance(z_array, np.ndarray))
+    def create_predictors(array_2):
+        if z_array is None:
+            return _create_predictors((array_2,))
+        return _create_predictors((array_2, z_array))
+
+    # Combine the various predictors
+    combined_obs_predictors = create_predictors(x2_array)
+
+    # Determine the number of rows being plotted
+    num_rows = x1_array.shape[0]
+
+    # Create a regressor to be used to compute the conditional expectations
+    regressor = _make_regressor(combined_obs_predictors, x1_array)
+
+    # Get the observed expectations
+    obs_expectation = regressor.predict(combined_obs_predictors)
+    obs_r2 = r2_score(x1_array, obs_expectation)
+
+    # Initialize arrays to store the permuted expectations and r2's
+    permuted_expectations = np.empty((num_rows, num_permutations))
+    permuted_r2 = np.empty(num_permutations, dtype=float)
+
+    # Get the permuted expectations
+    shuffled_index_array = np.arange(num_rows)
+
+    iterable = range(num_permutations)
+    if progress:
+        iterable = tqdm(iterable)
+
+    for i in iterable:
+        # Shuffle the index array
+        np.random.shuffle(shuffled_index_array)
+        # Get the new set of permuted X_2 values
+        current_x2 = x2_array[shuffled_index_array]
+        # Get the current combined predictors
+        current_predictors = create_predictors(current_x2)
+        # Fit a new model and store the current expectation
+        current_regressor =\
+            _make_regressor(current_predictors, x1_array)
+        permuted_expectations[:, i] =\
+            current_regressor.predict(current_predictors)
+        permuted_r2[i] = r2_score(x1_array, permuted_expectations[:, i])
+    return obs_r2, permuted_r2
+
+
+def visualize_permutation_results(obs_r2: float,
+                                  permuted_r2: np.ndarray,
+                                  verbose: bool=True,
+                                  permutation_color: str='#a6bddb',
+                                  output_path: Optional[str]=None,
+                                  show: bool=True,
+                                  close: bool=False) -> float:
+    """
+    Parameters
+    ----------
+    obs_r2 : float
```

Comment:
Is it necessary to put variable type even if you used typehinting?

Reply:
I did so simply to follow the [numpy documentation style](https://numpy.org/devdocs/docs/howto_document.html). It's unclear if it's strictly necessary anymore.

The benefit is with custom defined types (as example, see [here](https://github.com/hassanobeid1994/tr_b_causal_2020/blob/refactor-latent-cit/src/utils.py#L82)). If you hit tab in a jupyter notebook it will show the complete path to the underlying types which can be long and confusing to read. The docstring will just show whatever you write, which can be far simpler to understand.